### PR TITLE
Rest api for handle resolution with metadata

### LINKED ISSUE #760: prepare REST api for handle resolution with metadata
The new Handle server project ([link](https://github.com/dataquest-dev/docker-handle-server)) uses [this code](https://github.com/dataquest-dev/Remote-Handle-Resolver/blob/66caf4db85dddeb301ffb7e9132dd6efcf27ce6b/src/main/java/org/dspace/handle/MultiRemoteDSpaceRepositoryHandlePlugin.java#L249) or [this code](https://github.com/dataquest-dev/Remote-Handle-Resolver/blob/66caf4db85dddeb301ffb7e9132dd6efcf27ce6b/src/main/java/org/dspace/handle/MultiRemoteDSpaceRepositoryHandlePlugin.java#L309) to get a URL for a specific handle.

The endpoint "/resolve/" can be found in [HdlResolverRestController.java](https://github.com/dataquest-dev/DSpace/blob/446adde203bf0e71b1062642253eac5fe6376f3e/dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java#L96) and [here](https://github.com/dataquest-dev/DSpace/blob/446adde203bf0e71b1062642253eac5fe6376f3e/dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java#L174).

I want to modify this endpoint so that if a parameter is provided (or in some other way, such as updating the URL, HTTP request parameters, etc.), it will return a List Map, i.e., instead of https://dev-5.pc:8443/server/resolve/11234/1-5419 which returns ["AAA"], it should return [{"url": "AAA", OTHER METADATA}] where other metadata is like in http://hdl.handle.net/11025/57448?noredirect which is in [HandlePlugin.java](https://github.com/dataquest-dev/DSpace/blob/446adde203bf0e71b1062642253eac5fe6376f3e/dspace-api/src/main/java/org/dspace/handle/HandlePlugin.java#L608).

NOTE: (Note, you restarted Docker during the import and handle.remote-resolver.enabled = true is set to false, which is why /resolve/ does not work.)

SUMMARY: Instead of ["string"], I want /resolve/* [{}].

### PR DESCRIPTION
**Pull request by Paurikova2**
| Phases            | MP | MM  |  MB  | MR |   JM | Total  |
|-----------------|----:|----:|----:|-----:|-----:|-------:|
| ETA                  |  0  |  0  |    0 |     0 |      0 |        0 |
| Developing      |  10  |  0  |    0 |    0 |      0 |         0 |
| Review             |  0  |  0  |    0 |    0 |      0 |         0 |
| Total                |   -  |   -  |   -   |  -    |   -    |         0 |
| ETA est.             |      |      |       |       |         |         0 |
| ETA cust.           |   -  |   -  |   -  |   -   |   -     |        0 |
## Problem description
Return title, repository , submitdate and reportemail metadata for handle. not only url.

The new Handle server project ([link](https://github.com/dataquest-dev/docker-handle-server)) uses [this code](https://github.com/dataquest-dev/Remote-Handle-Resolver/blob/66caf4db85dddeb301ffb7e9132dd6efcf27ce6b/src/main/java/org/dspace/handle/MultiRemoteDSpaceRepositoryHandlePlugin.java#L249) or [this code](https://github.com/dataquest-dev/Remote-Handle-Resolver/blob/66caf4db85dddeb301ffb7e9132dd6efcf27ce6b/src/main/java/org/dspace/handle/MultiRemoteDSpaceRepositoryHandlePlugin.java#L309) to get a URL for a specific handle.

The endpoint "/resolve/" can be found in [HdlResolverRestController.java](https://github.com/dataquest-dev/DSpace/blob/446adde203bf0e71b1062642253eac5fe6376f3e/dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java#L96) and [here](https://github.com/dataquest-dev/DSpace/blob/446adde203bf0e71b1062642253eac5fe6376f3e/dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java#L174).

I want to modify this endpoint so that if a parameter is provided (or in some other way, such as updating the URL, HTTP request parameters, etc.), it will return a List Map, i.e., instead of https://dev-5.pc:8443/server/resolve/11234/1-5419 which returns ["AAA"], it should return [{"url": "AAA", OTHER METADATA}] where other metadata is like in http://hdl.handle.net/11025/57448?noredirect which is in [HandlePlugin.java](https://github.com/dataquest-dev/DSpace/blob/446adde203bf0e71b1062642253eac5fe6376f3e/dspace-api/src/main/java/org/dspace/handle/HandlePlugin.java#L608).

NOTE: (Note, you restarted Docker during the import and handle.remote-resolver.enabled = true is set to false, which is why /resolve/ does not work.)

SUMMARY: Instead of ["string"], I want /resolve/* [{}].

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/handle/HandlePlugin.java [hunk 1]
```diff
@@ -734,6 +752,10 @@ public byte[][] toRawValue() throws HandleException {
         return rawValues;
     }
 
+    public List<HandleValue> getHandleValue() {
```
Comment (milanmajchrak) [inline dspace-api/src/main/java/org/dspace/handle/HandlePlugin.java:755]: getHandleValues not one value

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java [hunk 2]
```diff
@@ -102,7 +110,7 @@ public ResponseEntity<String> resolveHandle(HttpServletRequest request, String h
         if (!handleResolver.isValid()) {
             return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
         } else {
-            return new ResponseEntity<>(this.resolveToURL(request, handleResolver), HttpStatus.OK);
+            return new ResponseEntity<>(this.resolveToMtd(request, handleResolver), HttpStatus.OK);
```
Comment (milanmajchrak) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:113]: what means Mtd?

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/hdlresolver/HdlResolverRestControllerIT.java [hunk 3]
```diff
@@ -76,17 +77,87 @@ public void givenMappedIdentifierWhenCallHdlresolverThenReturnsMappedURL() throw
...
+                    .param("repository", "true")
+                    .param("submitdate", "true")
+                    .param("reportemail", "true"))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$.title", StringContains.containsString("Public item 1")))
+                .andExpect(jsonPath("$.repository", is("DSpace at My University")))
+                .andExpect(jsonPath("$.url",
+                        StringContains.containsString("123456789/testHdlResolver")))
+                .andExpect(jsonPath("$.reportemail",
+                        StringContains.containsString("dspace-help@ufal.mff.cuni.cz")))
```
Comment (milanmajchrak) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/hdlresolver/HdlResolverRestControllerIT.java:124]: There should be generic report email, use testing `local.cfg`

**Review by milanmajchrak — CHANGES_REQUESTED** [review]

File: dspace-api/src/main/java/org/dspace/handle/HandlePlugin.java [hunk 4]
```diff
@@ -268,6 +268,24 @@ public void scanNAs(ScanCallback callback) throws HandleException {
...
+        ResolvedHandle rh = getHandleValues(theHandle, indexList, typeList);
+        if (Objects.isNull(rh)) {
+            return null;
+        }
+        return rh.getHandleValue();
+    }
+
+    private ResolvedHandle getHandleValues(byte[] theHandle, int[] indexList,
+                                           byte[][] typeList) throws HandleException {
         if (log.isInfoEnabled()) {
```
Comment (vidiecan) [inline dspace-api/src/main/java/org/dspace/handle/HandlePlugin.java:289]: add log to the interface methods (@Override)
fix names

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java [hunk 5]
```diff
@@ -171,14 +179,45 @@ public ResponseEntity<String> listHandles(HttpServletRequest request, @PathVaria
      * @param handleResolver HdlResolverDTO - Handle resolver
      * @return One element list using String if found, else null String.
      */
-    private String resolveToURL(HttpServletRequest request, HdlResolverDTO handleResolver) {
-        return mapAsJson(this.hdlResolverService.resolveToURL(ContextUtil.obtainContext(request), handleResolver));
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:175]: why was this changed to HandlePlugin()?

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java (hunk 2 above)
Comment (Paurikova2) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:113]: metadata

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java (hunk 5 above)
Comment (Paurikova2) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:175]: Because now the result should be not only the url, but also the reportemail, repository, etc. I wanted it to be more self-explanatory...

**Review by milanmajchrak — CHANGES_REQUESTED** [review]

**Review by milanmajchrak — DISMISSED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java [hunk 6]
```diff
@@ -165,14 +169,42 @@ public ResponseEntity<String> listHandles(HttpServletRequest request, @PathVaria
...
+
+    /**
+     * Resolves the metadata based on the given request and handle resolver.
+     *
+     * @param request        the HTTP request containing the metadata parameter
+     * @param handleResolver the handle resolver containing the handle
+     * @return a JSON representation of the URL or the handle values map, or "null" in case of an error
+     */
+    private String resolveToMtd(HttpServletRequest request, HdlResolverDTO handleResolver) {
+        Map<String, String> result = new HashMap<>();
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:190]: Declare variables as close as possible to their usage

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java [hunk 7]
```diff
@@ -165,14 +169,42 @@ public ResponseEntity<String> listHandles(HttpServletRequest request, @PathVaria
...
+        String param = request.getParameter("metadata");
+        if (StringUtils.isBlank(param)) {
+            return mapAsJson(url);
+        }
+        result.put("url", url);
+        String handle = handleResolver.getHandle();
+        try {
+            Map<String, String> metadata = HandlePlugin.getMapHandleValues(handle);
+            for (Map.Entry<String, String> entry : metadata.entrySet()) {
+                result.put(entry.getKey().toLowerCase(), entry.getValue());
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:201]: why lowercase?

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java [hunk 8]
```diff
@@ -102,7 +106,7 @@ public ResponseEntity<String> resolveHandle(HttpServletRequest request, String h
         if (!handleResolver.isValid()) {
             return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
         } else {
-            return new ResponseEntity<>(this.resolveToURL(request, handleResolver), HttpStatus.OK);
+            return new ResponseEntity<>(this.resolveToMtd(request, handleResolver), HttpStatus.OK);
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:109]: the name is not accurate, is it?

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/hdlresolver/HdlResolverRestControllerIT.java [hunk 9]
```diff
@@ -84,7 +85,37 @@ public void givenMappedIdentifierWhenCallHdlresolverThenReturnsMappedURL() throw
...
+        getClient()
+                .perform(get(HdlResolverRestController.RESOLVE  + publicItem1.getHandle())
+                    .param("metadata", "true"))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$.url",
+                        StringContains.containsString("123456789/testHdlResolver")))
+                .andExpect(jsonPath("$.title", StringContains.containsString("Public item 1")))
+                .andExpect(jsonPath("$.repository", is(configurationService.getProperty("dspace.name"))))
+                .andExpect(jsonPath("$.reportemail",
+                        StringContains.containsString(configurationService.getProperty("handle.reportemail"))))
```
[the remark this answers is not in the document]
Comment (Paurikova2) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/hdlresolver/HdlResolverRestControllerIT.java]: i don't understand...

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java (hunk 7 above)
Comment (Paurikova2) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:201]: I thought using lowercase might make it easier to use in the future, but it actually makes the code harder to read.

**Review by milanmajchrak — APPROVED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java [hunk 10]
```diff
@@ -166,13 +169,28 @@ public ResponseEntity<String> listHandles(HttpServletRequest request, @PathVaria
...
+        }
+        String handle = handleResolver.getHandle();
+        try {
+            Map<String, String> metadata = HandlePlugin.getMapHandleValues(handle);
+            metadata.put("URL", url);
+            return mapAsJson(metadata);
+        } catch (HandleException e) {
+            log.error("Failed to resolve handle values for handle: " + handle, e);
+        }
+        return "null";
```
Comment (vidiecan) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:193]: why "null"?

File: dspace-api/src/main/java/org/dspace/handle/HandlePlugin.java [hunk 11]
```diff
@@ -254,6 +254,54 @@ public void scanNAs(ScanCallback callback) throws HandleException {
...
+     */
+    public static Map<String, String> getMapHandleValues(String handle) throws HandleException {
+        if (log.isInfoEnabled()) {
+            log.info("Called getMapHandleValues");
+        }
+        loadServices();
+        Context context = new Context();
+        try {
+            DSpaceObject dso = null;
+            boolean resolveMetadata = configurationService.getBooleanProperty("lr.pid.resolvemetadata", true);
```
Comment (vidiecan) [inline dspace-api/src/main/java/org/dspace/handle/HandlePlugin.java:291]: something like ?
if (!resolveMetadata) {
 return null;
}

**Review by vidiecan — CHANGES_REQUESTED** [review]

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java (hunk 10 above)
Comment (Paurikova2) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/hdlresolver/HdlResolverRestController.java:193]: Because in original implementation it was like "null" -> https://github.com/dataquest-dev/DSpace/pull/761/files#diff-48aa4c8ed31e44c4791f50fbd34fb735f4164bf59bb77427ec20c545d150296eR307

File: dspace-api/src/main/java/org/dspace/handle/HandlePlugin.java (hunk 11 above)
Comment (Paurikova2) [reply dspace-api/src/main/java/org/dspace/handle/HandlePlugin.java:291]: But the resul of extraMetadata form null is empty map  not null https://github.com/dataquest-dev/DSpace/pull/761/files#diff-89c959e468673dbb5fad73c3dadd9ec90a76cf12f16066a989450e4d5ed85c76L572

Comment (vidiecan) [reply dspace-api/src/main/java/org/dspace/handle/HandlePlugin.java:291]: but on line 290 you init it with null

Comment (Paurikova2) [reply dspace-api/src/main/java/org/dspace/handle/HandlePlugin.java:291]: when dso is null, the return value is a `new LinkedHashMap<>()`, not `null`. But we will get this information in `extractMetadata`. Based on that, I believe it’s better to leave it as it is now.

**Review by vidiecan — APPROVED** [review]
